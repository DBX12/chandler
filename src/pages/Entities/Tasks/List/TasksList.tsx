import React, {PureComponent} from 'react';
import {ActivityIndicator, FlatList, Text, View} from 'react-native';

import {Checkbox, EmptyActivity, Navbar} from '@components';
import {I18n} from '@i18n';
import {ITask} from '@models';
import {IPopAction} from '@navigator/NavigationService';
import {commonStyles} from '@theme';
import {styles} from './TasksList.styles';

interface ITasksListProps {
  pop: IPopAction;
  getTasksByContact: () => void;
  tasks: ITask[];
  isFetching: boolean;
}

export class TasksList extends PureComponent<ITasksListProps, {}> {
  public componentWillMount() {
    this.props.getTasksByContact();
  }

  public keyExtractor = (item, index) => String(item.id);

  public renderFooter = () => {
    const {isFetching} = this.props;

    if (!isFetching) {
      return null;
    }

    return (
      <ActivityIndicator style={commonStyles.activityIndicator} size="large" />
    );
  };

  public renderItem = ({item, index}) => {
    const {tasks} = this.props;
    const task = tasks[index];

    return (
      <View style={styles.taskContainer}>
        <Checkbox checked={task.completed} />
        <View style={styles.textContainer}>
          <Text>{task.title}</Text>
          {task.description && (
            <Text style={styles.descriptionText}>{task.description}</Text>
          )}
        </View>
      </View>
    );
  };

  public render() {
    const {pop, tasks, getTasksByContact, isFetching} = this.props;

    return (
      <View style={commonStyles.flex}>
        <Navbar title={I18n.t('tasks:tasks')} onBack={pop} />
        {isFetching || tasks.length ? (
          <FlatList
            data={tasks}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            ListFooterComponent={this.renderFooter}
            onEndReached={getTasksByContact}
            onEndReachedThreshold={0.5}
          />
        ) : (
          <EmptyActivity
            image={require('./assets/empty-tasks.png')}
            title={I18n.t('tasks:emptyTitle')}
            subtitle={I18n.t('tasks:emptySubtitle')}
          />
        )}
      </View>
    );
  }
}
