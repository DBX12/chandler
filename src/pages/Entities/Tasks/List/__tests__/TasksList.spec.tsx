import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import {TasksList} from '../TasksList';

describe('Pages', () => {
  describe('Tasks', () => {
    let pop;
    let getTasksByContact;

    beforeEach(() => {
      pop = jest.fn();
      getTasksByContact = jest.fn();
    });

    it('should renders correctly without tasks and fetching', () => {
      const tasks = [];
      const tree = shallow(
        <TasksList
          pop={pop}
          getTasksByContact={getTasksByContact}
          isFetching={true}
          tasks={tasks}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly without tasks and not fetching', () => {
      const tasks = [];
      const tree = shallow(
        <TasksList
          pop={pop}
          getTasksByContact={getTasksByContact}
          isFetching={false}
          tasks={tasks}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly with tasks and fetching', () => {
      const tasks = [{title: 'My title', completed: false}];
      const tree = shallow(
        <TasksList
          pop={pop}
          getTasksByContact={getTasksByContact}
          isFetching={true}
          tasks={tasks as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
    });

    it('should renders correctly with tasks and fetching', () => {
      const tasks = [
        {title: 'My title', completed: false},
        {title: 'My title 2', completed: true, description: 'My desc'},
      ];
      const tree = shallow(
        <TasksList
          pop={pop}
          getTasksByContact={getTasksByContact}
          isFetching={false}
          tasks={tasks as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 1})))
      ).toMatchSnapshot();
    });

    it('should display the task indicator in the footer if it is fetching', () => {
      const tasks = [
        {title: 'My title', completed: false},
        {title: 'My title 2', completed: false, description: 'My desc'},
      ];
      const tree = shallow(
        <TasksList
          pop={pop}
          getTasksByContact={getTasksByContact}
          isFetching={true}
          tasks={tasks as any}
        />
      );
      expect(toJson(shallow(tree.instance().renderFooter()))).toMatchSnapshot();
    });

    it('should not display the task indicator in the footer if it is not fetching', () => {
      const tasks = [
        {title: 'My title', completed: false},
        {title: 'My title 2', completed: true, description: 'My desc'},
      ];
      const tree = shallow(
        <TasksList
          pop={pop}
          getTasksByContact={getTasksByContact}
          isFetching={false}
          tasks={tasks as any}
        />
      );
      expect(tree.instance().renderFooter()).toBeNull();
    });

    it('should have a keyExtractor', () => {
      const tasks = [{id: 1}, {id: 2}];
      const tree = shallow(
        <TasksList
          pop={pop}
          getTasksByContact={getTasksByContact}
          isFetching={false}
          tasks={tasks as any}
        />
      );
      expect(tree.instance().keyExtractor(tasks[0])).toBe('1');
      expect(tree.instance().keyExtractor(tasks[1])).toBe('2');
    });
  });
});
