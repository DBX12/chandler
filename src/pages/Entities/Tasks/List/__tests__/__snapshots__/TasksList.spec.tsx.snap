// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Tasks should display the task indicator in the footer if it is fetching 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="large"
  style={
    Object {
      "marginVertical": 20,
    }
  }
/>
`;

exports[`Pages Tasks should renders correctly with tasks and fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Tasks"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "completed": false,
          "title": "My title",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages Tasks should renders correctly with tasks and fetching 2`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "flexDirection": "row",
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
>
  <Checkbox
    checked={false}
  />
  <Component
    style={
      Object {
        "paddingLeft": 15,
        "paddingRight": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      My title
    </Text>
  </Component>
</View>
`;

exports[`Pages Tasks should renders correctly with tasks and fetching 3`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Tasks"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "completed": false,
          "title": "My title",
        },
        Object {
          "completed": true,
          "description": "My desc",
          "title": "My title 2",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages Tasks should renders correctly with tasks and fetching 4`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "flexDirection": "row",
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
>
  <Checkbox
    checked={false}
  />
  <Component
    style={
      Object {
        "paddingLeft": 15,
        "paddingRight": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      My title
    </Text>
  </Component>
</View>
`;

exports[`Pages Tasks should renders correctly with tasks and fetching 5`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "flexDirection": "row",
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
>
  <Checkbox
    checked={true}
  />
  <Component
    style={
      Object {
        "paddingLeft": 15,
        "paddingRight": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      My title 2
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#818181",
          "fontSize": 14,
          "marginTop": 5,
        }
      }
    >
      My desc
    </Text>
  </Component>
</View>
`;

exports[`Pages Tasks should renders correctly without tasks and fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Tasks"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages Tasks should renders correctly without tasks and not fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Tasks"
  />
  <EmptyActivity
    image={1}
    subtitle="Let Monica manage stuff for you."
    title="You have no excuses not to be multitask anymore."
  />
</Component>
`;
